### PR TITLE
Ensure that buffer used for compressing message is released

### PR DESCRIPTION
**Pull request by jyemin**
JAVA-3087

Patch build: https://evergreen.mongodb.com/version/5be9b6dfe3c331337868de21

### CODE REVIEW

**Comment by jyemin** [comment]
Hold off on this one.  Patch build has failed.  Looks like there was a reason that the buffer wasn't closed until later.

**Comment by jyemin** [comment]
New patch build: https://evergreen.mongodb.com/version/5be9d26fe3c3313378697168

File: driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java [hunk 1]
```diff
@@ -264,23 +264,29 @@ public boolean isClosed() {
...
+            try {
                 sendMessage(bsonOutput.getByteBuffers(), message.getId());
-            } else {
-                CompressedMessage compressedMessage = new CompressedMessage(message.getOpCode(), bsonOutput.getByteBuffers(),
-                        sendCompressor,
+            } finally {
+                bsonOutput.close();
+            }
+        } else {
+            List<ByteBuf> byteBuffers = bsonOutput.getByteBuffers();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java]: Fixing it in synchronous path as well.

Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java]: The underlying issues is that #getByteBuffers duplicates the buffers, which requires that they be released independently of the BsonOutput that they came from.

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java]: So compression causes a double copy of buffers?

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java]: Two buffers, but different contents.  The first is the uncompressed message.  The second is the compressed message.

**Pull request closed without merging**